### PR TITLE
List available packages if providing `--package` with an empty value

### LINKED ISSUE #8591: cargo run -p should list valid packages like --bin lists targets
**Describe the problem you are trying to solve**
We have a virtual workspace with multiple sub-projects and _tools_ (a workspace containing an app we need during development). Before introducing features we got used to running `cargo run --bin`, which would then print:

```
error: "--bin" takes one argument.
Available binaries:
    tool_a
    tool_b
    tool_c
    ...
```
Since tools get refactored a lot it was a nice way of seeing all options without too much ceremony.  However, after adding features, running `cargo run --bin xxx` does not reliably produce the desired target anymore, but `cargo run -p xxx` does, so we tried switching to that one. 

However, running `cargo bin -p` gives:

```
error: The argument '--package <SPEC>' requires a value but none was supplied

USAGE:
    cargo.exe run --package <SPEC>

For more information try --help
```

**Describe the solution you'd like**
It would be nice if `cargo bin -p` lists all projects it can currently run, similar to how `--bin` works. 

**Notes**
I don't know if `-p` is more powerful than `--bin` in the sense that it accepts binaries outside the virtual workspace. In that case it would be nice to still list all local ones, and maybe note that other targets could be available.

### PR DESCRIPTION
**Pull request by weihanglo**
May resolves #8591 

## How

First we need to take the responsibility of check command line arguments from claps. I've examine all 10 build commands and all of them call [`ArgMatchesExt::compile_options`](https://github.com/rust-lang/cargo/blob/2f115a76e5a1e5eb11cd29e95f972ed107267847/src/cargo/util/command_prelude.rs#L389-L395) directly or indirectly. And `compile_options` [calls `check_optional_opts`](https://github.com/rust-lang/cargo/blob/2f115a76e5a1e5eb11cd29e95f972ed107267847/src/cargo/util/command_prelude.rs#L499-L501) to check if target selection options given an empty value. So we can do the same logic there.

I've also add a error message for an edge case though that one would never trigger at this moment.

### CODE REVIEW

**Comment by Eh2406** [comment]
This looks good to me,
@bors r? @ehuss for a second opinion.

File: src/cargo/util/workspace.rs [hunk 1]
```diff
@@ -38,29 +39,50 @@ fn print_available(
     let mut output = String::new();
     writeln!(output, "\"{}\" takes one argument.", option_name)?;
 
-    if targets.is_empty() {
+    print_availables(output, &targets, plural_name)
+}
+
+fn print_availables(output: String, availables: &[&str], plural_name: &str) -> CargoResult<()> {
+    let mut output = output;
```
Comment (ehuss) [inline src/cargo/util/workspace.rs:46]:
```suggestion
fn print_availables(mut output: String, availables: &[&str], plural_name: &str) -> CargoResult<()> {
```

File: src/cargo/util/workspace.rs [hunk 2]
```diff
@@ -38,29 +39,50 @@ fn print_available(
...
+        .members()
+        .map(|pkg| pkg.name().as_str())
+        .collect::<Vec<_>>();
+
+    let mut output = String::new();
+    writeln!(
+        output,
+        "\"--package <SPEC>\" requires a SPEC format value.\n\
+        Run `cargo help pkgid` for more infomation about SPEC format."
+    )?;
```
Comment (ehuss) [inline src/cargo/util/workspace.rs:69]:
```suggestion
    let output = "\"--package <SPEC>\" requires a SPEC format value.\n\
        Run `cargo help pkgid` for more infomation about SPEC format.\n"
        .to_string();
```

**Review by ehuss — COMMENTED** [review]
Thanks!  Looks useful!

I noticed that some commands do not seem to correctly validate the argument with this change. `clean`, `tree`, `uninstall`, and `update` will silently run with a `-p` argument without a value, when they should return an error. Can you adjust those so that they will fail (and preferably print out an available list)?

I also noticed that the single-argument commands (`rustc`, `rustdoc`, `pkgid`, and `run`) don't print the available list. I won't block the PR if you don't want to add them, but I think it would be nice.

Finally, I'm a little concerned that this might contribute some confusion. In most cases, the `-p` argument can accept any package in the resolve graph, not just workspace members. However, I think listing everything would be too noisy. I'm not sure how to improve that. Maybe if it said "Available workspace members:" or something like that?

**Comment by weihanglo** [comment]
> I also noticed that the single-argument commands (`rustc`, `rustdoc`, `pkgid`, and `run`) don't print the available list. I won't block the PR if you don't want to add them, but I think it would be nice.

Oh I see. I was thinking that the only accept one package so showing a list does not help. But I was wrong. It still benefits  😂. Will fix it.

> I'm not sure how to improve that. Maybe if it said "Available workspace members:" or something like that?

That’s also what I was concerned about, therefore I decided to add a help message guiding people to learn more about pkgid. Showing "Available workspace members:" seems very straightforward. However, if a single-package crate shows message like that, would it confuses user about workspace/package, though both kind of compilations can get workspace information internally in cargo.

Would this be better?

```
[ERROR] "--package <SPEC>" requires a SPEC format value, which can be any package ID specifier in the dependency graph.
Run `cargo help pkgid` for more information about SPEC format."

Possible packages:
    foo

**Comment by ehuss** [comment]
Looks good, thanks!
@bors r+

**Pull request merged**